Initial support for Alif Semiconductor Ensemble SoC family and Devkit

silesh-alifsemi opened the pull request:
Ensemble is a family of processors from Alif Semiconductor targeted at low-end to high-end intelligent IoT endpoint
applications. Series (E1-E8) within the family scale up from single-core MCUs to quad-core MCU/MPU fusion processors.

https://alifsemi.com/products/ensemble/
https://alifsemi.com/products/ensemblegenai/

The E8 series within the Ensemble family features two Cortex-M55 cores, two Cortex-A32 cores, Ethos-U85 and dual Ethos-U55 NPUs.

 https://alifsemi.com/ensemble-e8-series/

This PR adds basic support for the Alif Semiconductor Ensemble E8 development kit featuring the AE822FA0E5597LS0 SoC from the E8 series.

https://alifsemi.com/support/kits/ensemble-e8devkit

Initial support includes:
- Support for the Ensemble SOC family, E8 SoC series and AE822FA0E5597LS0 SoC
- Support for the two Cortex-M55 CPU clusters (Real time subsystems: RTSS-HP and RTSS-HE)
- Clock control driver
- Pinctrl driver
- UART support
- Support for the Ensemble E8 development kit

silesh-alifsemi committed:
dts: arm: alif: ensemble: add RTSS-HE and RTSS-HP subsystems

The Alif Ensemble family of processors contains up to two
Cortex-M55 cores, each optimized for different use cases. One
configured to maximize power efficiency and the other to maximize
compute performance. These are called the High Efficiency Real Time
Subsystem (RTSS-HE) and the High Performance Real Time Subsystem
(RTSS-HP) respectively.

Add the device tree include files for these subsystems with their
CPU, MPU and TCM definitions.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: arm: alif: ensemble: add peripheral mpu region

The Ensemble SoCs have a peripheral region at 0x1A000000 (8 MB)
that hosts peripherals such as pinmux and clock control registers.
This deviates from the standard Cortex-M memory map and requires
explicit MPU configuration with device memory attributes.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: bindings: vendor-prefixes: add alif prefix

Add prefix for Alif Semiconductor.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
soc: alif: introduce Ensemble family and E8 series

Ensemble is a family of processors from Alif Semiconductor
targeted at low-end to high-end intelligent IoT endpoint
applications. Series (E1-E8) within the family scale up from
single-core MCUs to quad-core MCU/MPU fusion processors.

Add support for the Ensemble family and the E8 series with the
AE822FA0E5597LS0 SoC. The SoC contains two Cortex-M55 cores: one
configured to maximize power efficiency (RTSS-HE - High Efficiency
Real Time Subsystem) and the other to maximize compute performance
(RTSS-HP - High Performance Real Time Subsystem).

Link: https://alifsemi.com/products/ensemble/
Link: https://alifsemi.com/products/ensemblegenai/
Link: https://alifsemi.com/ensemble-e8-series/

Co-authored-by: Tanay Rami <tanay@alifsemi.com>
Co-authored-by: Prabhakar Kumar <prabhakar.kumar@alifsemi.com>
Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: bindings: pinctrl: add alif pinctrl bindings

Add device tree bindings for the Alif Semiconductor pin controller
and Ensemble SoC specific pin definitions.

The binding supports pin multiplexing via predefined macros that
encode port, pin, and alternate function into a 32-bit value.
Pad configuration (input enable, drive strength, slew rate, bias
control, etc.) is specified using standard pinctrl properties.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
drivers: pinctrl: add alif pinctrl driver

Add pin control driver for Alif SoCs. The driver manages
pin muxing (alternate function selection) and pad configuration
for Alif SoCs.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: arm: alif: ensemble: Add pinctrl node

Add pinctrl device tree node for Alif Ensemble SoC with main and
LPGPIO pinmux register regions.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: arm: alif: ensemble: add AE822FA0E5597LS0 SoC dtsi

Add device tree include file for the Alif Ensemble E8 SoC
AE822FA0E5597LS0.

This dtsi defines resources specific to this SoC variant in
addition to the common peripheral definitions from ensemble_common.dtsi.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
boards: alif: add Alif Ensemble E8 Development Kit

Add board support for the Alif Ensemble E8 Development Kit
featuring the Enemble E8 SoC.

The Ensemble E8 SoC contains two Cortex-M55 CPU clusters: RTSS-HP
(High Performance, up to 400 MHz) and RTSS-HE (High Efficiency,
up to 160 MHz). Following the hardware model, each CPU cluster
is defined as a separate board target:

- ensemble_e8_dk/ae822fa0e5597ls0/rtss_hp
- ensemble_e8_dk/ae822fa0e5597ls0/rtss_he

Initial support includes UART for console output.

Link: https://alifsemi.com/support/kits/ensemble-e8devkit

Co-authored-by: Tanay Rami <tanay@alifsemi.com>
Co-authored-by: Prabhakar Kumar <prabhakar.kumar@alifsemi.com>
Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dt-bindings: clock: add alif clock bindings

Alif SoCs organize clock control into multiple register regions
(called clock modules in the driver). Bitfields in these modules
are used to manage different clocks in the SoC. Add headers that
define the encoding of the information necessary to operate on
the clocks.

The encoding packs clock control information into a single 32-bit
value that includes clock module and register offset within the
module, clock enable information and clock source information.

Note that alif-clocks-common.h contains the common encoding scheme,
module definitions etc. that could be reused by other Alif SoCs and
alif-ensemble-clocks.h defines the peripheral clock IDs specific to
the Ensemble family (UART 0-7 clocks are added for initial support).

Co-authored-by: Manoj A Murudi <manoj.murudi@alifsemi.com>
Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: arm: alif: ensemble: add fixed root clocks

Add fixed-clock nodes for some of the Ensemble family root clocks.
These clocks serve as input sources for peripheral clocks managed
by the clock controller.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: bindings: clock: add alif clock controller binding

Add devicetree binding for the Alif Semiconductor clock controller.
The clock controller manages peripheral clocks using registers spread
across multiple discontiguous regions (called clock modules in the
binding and driver).

Clock specifiers use a single cell containing an encoded clock ID
that packs module, register offset, clock enable information and
source clock information.

Co-authored-by: Manoj A Murudi <manoj.murudi@alifsemi.com>
Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
dts: arm: alif: ensemble: add clock controller node

Add the clock controller node for the Ensemble family.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
drivers: clock_control: add alif clock driver

Add clock control driver for Alif Semiconductor SoCs. The driver
manages peripheral clocks in the device by operating on multiple
discontiguous register regions (called modules in this driver).
In this initial version, the driver supports on, off, get_rate and
get_status API.

Co-authored-by: Manoj A Murudi <manoj.murudi@alifsemi.com>
Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi committed:
MAINTAINERS: add alif semiconductor entries

Add maintainers and collaborators for Alif Semiconductor
platforms.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

JarmouniA commented on soc/alif/ensemble/common/mpu_regions.c:
```diff
@@ -0,0 +1,30 @@
+	MPU_REGION_ENTRY("FLASH", CONFIG_FLASH_BASE_ADDRESS,
+			REGION_FLASH_ATTR(CONFIG_FLASH_BASE_ADDRESS, CONFIG_FLASH_SIZE * 1024)),
+
+	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
```
ditto
https://github.com/zephyrproject-rtos/zephyr/blob/9a0e8a4148ceb6e3dc223c5f854eb19472843d90/arch/arm/core/mpu/arm_mpu_regions.c#L32

silesh-alifsemi replied:
Moved the peripheral region definition into a device tree node.

silesh-alifsemi committed:
dts: arm: alif: ensemble: add uart nodes

Add device tree nodes for ns16550 compatible uart instances common
to all Socs in the ensemble family. Additional uart instances available
on specific SoCs will be added in the corresponding SoC specific dtsi
file.

Signed-off-by: Silesh C V <silesh@alifsemi.com>

silesh-alifsemi commented on soc/alif/ensemble/common/mpu_regions.c:
```diff
@@ -0,0 +1,30 @@
+	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
+			REGION_RAM_ATTR(CONFIG_SRAM_BASE_ADDRESS, CONFIG_SRAM_SIZE * 1024)),
+
+	MPU_REGION_ENTRY("PERIPHERALS", PERIPHERAL_BASE,
```
Done. Moved the peripheral region definition into a device tree node.

silesh-alifsemi commented on soc/alif/ensemble/e8/Kconfig.defconfig.ae822fa0e5597ls0_rtss_hp:
```diff
@@ -0,0 +1,12 @@
+	default 480
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 400000000
```
Fixed.

JarmouniA commented on dts/bindings/pinctrl/alif,pinctrl.yaml:
```diff
@@ -0,0 +1,131 @@
+# Copyright (c) 2025 Alif Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+title: Alif Semiconductor pin controller
```
```suggestion
title: Pin controller
```
The vendor name is visible in the compatible string.
This will make the name shorter in the generated Board Features table in the doc.

silesh-alifsemi commented on soc/alif/ensemble/e8/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2025 Alif Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_E8_RTSS_COMMON
```
Agreed. Renamed to SOC_SERIES_E8_RTSS_COMMON. I have kept the _COMMON suffix to indicate that this helper symbol groups all the common architecture configurations that are specific to the RTSS cores in the E8 series SoCs.

VeijoPesonen replied:
@nordicjm So is SOC_SERIES_E8_RTSS_COMMON fine for you based on the comments from @silesh-alifsemi ?

silesh-alifsemi replied:
Fixed. Removed the `_COMMON` suffix from the config symbol name.

kartben reviewed (dismissed):
+1 for docs - excite to see this coming!

kartben commented on dts/bindings/pinctrl/alif,pinctrl.yaml:
```diff
@@ -0,0 +1,131 @@
+# Copyright (c) 2025 Alif Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+title: Alif pin controller
```
for  consistency with the clock controller binding

```suggestion
title: Alif Semiconductor pin controller
```

silesh-alifsemi replied:
Done.

silesh-alifsemi commented on dts/bindings/clock/alif,clockctrl.yaml:
```diff
@@ -0,0 +1,48 @@
+  Example usage in a peripheral node:
+    uart0: uart@... {
+        clocks = <&clockctrl ALIF_UART0_SYST_PCLK>;
+    };
```
Done.

kartben commented:
> removing assignee as clock control is really only a small subset of what this PR is about :)

Removed again for the same reason.
